### PR TITLE
Performance Testing Framework for Kokkos (Implementing RAJAPerfSuite Infrastructure using git submodules)

### PR DESCRIPTION
**Pull request by ajpowelsnl**
We have created an initial implementation of the perf testing framework for Kokkos.  CAVEAT:  this does not represent a final, polished state for the implementation, but rather a first draft to communicate and discuss the concept, and elicit feedback.

Here, we are brining RAJAPerf Suite in via git submodules as a Kokkos tpl (in the same manner as we've done for Kokkos Kernels -- https://github.com/kokkos/kokkos-kernels/pull/1037), and have provided a new sample test to run with this infrastructure, ViewAllocationPerfTest.

Here are some helpful hints to run this code on kokkos-dev-2:

module list
Currently Loaded Modulefiles:
  1) sems-env             3) sems-cmake/3.19.1    5) sems-gcc/9.2.0       7) /clang/12.0.0_rc3
  2) sems-cuda/11.1       4) sems-openmpi/4.0.2   6) sems-python/3.5.2

_**IN A NEW KOKKOS PROJECT**_

cd kokkos/

git clone --recursive

_**IN AN EXISTING KOKKOS PROJECT**_

cd kokkos/

git submodule update --init --recursive

**_BUILD TO VERIFY BEHAVIOR_**

in kokkos:

mkdir build_submod

cd build_submod

../cmake \
-DKokkos_ENABLE_TRACKED_TESTING=ON \
-DKokkos_ENABLE_TESTS=ON \
.. \

- Assuming successful configure and build, change directory to the "tracked_testing" executable:

cd tpls/rajaperf/bin/

- In this directory , you will see:

ls -1
KokkosAlgorithms_UnitTest
KokkosContainers_UnitTest_Serial
KokkosCore_IncrementalTest_SERIAL
KokkosCore_PerformanceTest_Atomic
KokkosCore_PerformanceTest_Atomic_MinMax
KokkosCore_PerformanceTest_Mempool
KokkosCore_PerformanceTest_TaskDag
KokkosCore_PerfTestExec
KokkosCore_ProfilingAllCalls
KokkosCore_push_finalize_hook_terminate
KokkosCore_StackTraceTestExec
KokkosCore_tracked_testing
KokkosCore_UnitTest_CMakePassCmdLineArgs
KokkosCore_UnitTest_CTestDevice
KokkosCore_UnitTest_Default
KokkosCore_UnitTest_DefaultInit_1
KokkosCore_UnitTest_DefaultInit_10
KokkosCore_UnitTest_DefaultInit_11
KokkosCore_UnitTest_DefaultInit_12
KokkosCore_UnitTest_DefaultInit_13
KokkosCore_UnitTest_DefaultInit_14
KokkosCore_UnitTest_DefaultInit_15
KokkosCore_UnitTest_DefaultInit_16
KokkosCore_UnitTest_DefaultInit_17
KokkosCore_UnitTest_DefaultInit_18
KokkosCore_UnitTest_DefaultInit_2
KokkosCore_UnitTest_DefaultInit_3
KokkosCore_UnitTest_DefaultInit_4
KokkosCore_UnitTest_DefaultInit_5
KokkosCore_UnitTest_DefaultInit_6
KokkosCore_UnitTest_DefaultInit_7
KokkosCore_UnitTest_DefaultInit_8
KokkosCore_UnitTest_DefaultInit_9
KokkosCore_UnitTest_Develop
KokkosCore_UnitTest_HostBarrier
KokkosCore_UnitTest_LogicalSpaces
KokkosCore_UnitTest_PushFinalizeHook
KokkosCore_UnitTest_Serial1
KokkosCore_UnitTest_Serial2
KokkosCore_UnitTest_SerialGraph
Kokkos_HeaderSelfContained_Kokkos_AcquireUniqueTokenImpl_hpp
Kokkos_HeaderSelfContained_Kokkos_AnonymousSpace_hpp
Kokkos_HeaderSelfContained_Kokkos_Array_hpp
Kokkos_HeaderSelfContained_Kokkos_Atomic_hpp
Kokkos_HeaderSelfContained_Kokkos_Bitset_hpp
Kokkos_HeaderSelfContained_Kokkos_Complex_hpp
Kokkos_HeaderSelfContained_Kokkos_Concepts_hpp
Kokkos_HeaderSelfContained_Kokkos_CopyViews_hpp
Kokkos_HeaderSelfContained_Kokkos_Core_fwd_hpp
Kokkos_HeaderSelfContained_Kokkos_Core_hpp
Kokkos_HeaderSelfContained_Kokkos_Crs_hpp
Kokkos_HeaderSelfContained_Kokkos_Cuda_hpp
Kokkos_HeaderSelfContained_Kokkos_CudaSpace_hpp
Kokkos_HeaderSelfContained_Kokkos_DualView_hpp
Kokkos_HeaderSelfContained_Kokkos_DynamicView_hpp
Kokkos_HeaderSelfContained_Kokkos_DynRankView_hpp
Kokkos_HeaderSelfContained_Kokkos_ErrorReporter_hpp
Kokkos_HeaderSelfContained_Kokkos_ExecPolicy_hpp
Kokkos_HeaderSelfContained_KokkosExp_MDRangePolicy_hpp
Kokkos_HeaderSelfContained_Kokkos_Extents_hpp
Kokkos_HeaderSelfContained_Kokkos_Functional_hpp
Kokkos_HeaderSelfContained_Kokkos_Future_hpp
Kokkos_HeaderSelfContained_Kokkos_Graph_fwd_hpp
Kokkos_HeaderSelfContained_Kokkos_Graph_hpp
Kokkos_HeaderSelfContained_Kokkos_GraphNode_hpp
Kokkos_HeaderSelfContained_Kokkos_Half_hpp
Kokkos_HeaderSelfContained_Kokkos_HBWSpace_hpp
Kokkos_HeaderSelfContained_Kokkos_HIP_hpp
Kokkos_HeaderSelfContained_Kokkos_HIP_Space_hpp
Kokkos_HeaderSelfContained_Kokkos_HostSpace_hpp
Kokkos_HeaderSelfContained_Kokkos_HPX_hpp
Kokkos_HeaderSelfContained_Kokkos_hwloc_hpp
Kokkos_HeaderSelfContained_Kokkos_Layout_hpp
Kokkos_HeaderSelfContained_Kokkos_LogicalSpaces_hpp
Kokkos_HeaderSelfContained_Kokkos_Macros_hpp
Kokkos_HeaderSelfContained_Kokkos_MasterLock_hpp
Kokkos_HeaderSelfContained_Kokkos_MemoryPool_hpp
Kokkos_HeaderSelfContained_Kokkos_MemoryTraits_hpp
Kokkos_HeaderSelfContained_Kokkos_NumericTraits_hpp
Kokkos_HeaderSelfContained_Kokkos_OffsetView_hpp
Kokkos_HeaderSelfContained_Kokkos_OpenMP_hpp
Kokkos_HeaderSelfContained_Kokkos_OpenMPTarget_hpp
Kokkos_HeaderSelfContained_Kokkos_OpenMPTargetSpace_hpp
Kokkos_HeaderSelfContained_Kokkos_Pair_hpp
Kokkos_HeaderSelfContained_Kokkos_Parallel_hpp
Kokkos_HeaderSelfContained_Kokkos_Parallel_Reduce_hpp
Kokkos_HeaderSelfContained_Kokkos_PointerOwnership_hpp
Kokkos_HeaderSelfContained_Kokkos_Profiling_ProfileSection_hpp
Kokkos_HeaderSelfContained_Kokkos_Random_hpp
Kokkos_HeaderSelfContained_Kokkos_ScatterView_hpp
Kokkos_HeaderSelfContained_Kokkos_ScratchSpace_hpp
Kokkos_HeaderSelfContained_Kokkos_Serial_hpp
Kokkos_HeaderSelfContained_Kokkos_Sort_hpp
Kokkos_HeaderSelfContained_Kokkos_StaticCrsGraph_hpp
Kokkos_HeaderSelfContained_Kokkos_SYCL_hpp
Kokkos_HeaderSelfContained_Kokkos_SYCL_Space_hpp
Kokkos_HeaderSelfContained_Kokkos_TaskPolicy_hpp
Kokkos_HeaderSelfContained_Kokkos_TaskScheduler_fwd_hpp
Kokkos_HeaderSelfContained_Kokkos_TaskScheduler_hpp
Kokkos_HeaderSelfContained_Kokkos_Threads_hpp
Kokkos_HeaderSelfContained_Kokkos_Timer_hpp
Kokkos_HeaderSelfContained_Kokkos_Tuners_hpp
Kokkos_HeaderSelfContained_Kokkos_UniqueToken_hpp
Kokkos_HeaderSelfContained_Kokkos_UnorderedMap_hpp
Kokkos_HeaderSelfContained_Kokkos_Vector_hpp
Kokkos_HeaderSelfContained_Kokkos_Vectorization_hpp
Kokkos_HeaderSelfContained_Kokkos_View_hpp
Kokkos_HeaderSelfContained_Kokkos_WorkGraphPolicy_hpp

- run this executable (in kokkos/$BUILD_DIR/tpls/rajaperf/bin:

./KokkosCore_tracked_testing 

Reading command line input...

Reading command line input...

Setting up suite based on input...

RAJA performance suite run summary....
--------------------------------------

How suite will be run:
	 # passes = 1
	 Kernel size factor = 1
	 Kernel rep factor = 1
	 Output files will be named ./RAJAPerf*

The following kernels and variants (when available) will be run:

Variants
--------
Base_Seq
Lambda_Seq
RAJA_Seq
Kokkos_Lambda
Kokkos_Functor

Kernels(iterations/rep , reps)
-----------------------------
ViewAllocationTest (100000 , 50)

Run warmup kernel...

Running specified kernels and variants...

Generate run report files...

DONE!!!....

- Expected output:

RAJAPerf-checksum.txt
RAJAPerf-fom.csv
RAJAPerf-speedup.csv
RAJAPerf-timing.csv

- In RAJAPerf-timing.csv, you will see somethign on the order of:

Mean Runtime Report (sec.)  ,  ,  ,  ,  ,
Kernel              , Base_Seq , Lambda_Seq , RAJA_Seq , Kokkos_Lambda , Kokkos_Functor
ViewAllocationTest  ,  Not run ,    Not run ,  Not run ,      2.929409 ,        Not run

- We will clean up the executable, and location where the tests are run.  Again, this PR represents approximate behavior and output.
- If desired, we can put this PR on the Wed. meeting agenda for in depth discussion.

### CODE REVIEW

**Comment by masterleinad** [comment]
It seems there are a lot of unrelated commits in the commit history. Can you please rebase on top of `develop`?

File: cmake/kokkos_tribits.cmake [hunk 1]
```diff
@@ -79,6 +79,12 @@ MACRO(KOKKOS_PROCESS_SUBPACKAGES)
   if (KOKKOS_HAS_TRILINOS)
     TRIBITS_PROCESS_SUBPACKAGES()
   else()
+  # AJP & DZP added for RPS - based perf testing
```
Comment (masterleinad) [inline cmake/kokkos_tribits.cmake:82]: This sounds pretty cryptic. IMHO, we don't need another comment here.
```suggestion
```

File: cmake/kokkos_tribits.cmake [hunk 2]
```diff
@@ -79,6 +79,12 @@ MACRO(KOKKOS_PROCESS_SUBPACKAGES)
...
+  if (KOKKOS_ENABLE_TRACKED_TESTING)
+    SET(INFRASTRUCTURE_ONLY ON CACHE BOOL "Only build the RAJAPerf infrastructure,no builtin kernels")
```
Comment (masterleinad) [inline cmake/kokkos_tribits.cmake:84]: Why does this variable need to be cached?
```suggestion
    SET(INFRASTRUCTURE_ONLY ON CACHE BOOL "Only build the RAJAPerf infrastructure, no builtin kernels")
```

Also, what does this mean? Are all normal unit tests not run when `RAJAPerf` is enabled to be run?

File: cmake/kokkos_tribits.cmake [hunk 3]
```diff
@@ -79,6 +79,12 @@ MACRO(KOKKOS_PROCESS_SUBPACKAGES)
...
+    ADD_DEFINITIONS("-DRAJAPERF_INFRASTRUCTURE_ONLY")
```
Comment (masterleinad) [inline cmake/kokkos_tribits.cmake:85]: Is there a good reason not to use preprocessor variables defined by `#cmakedefine` as we usually do?

File: cmake/kokkos_tribits.cmake [hunk 4]
```diff
@@ -79,6 +79,12 @@ MACRO(KOKKOS_PROCESS_SUBPACKAGES)
...
+	ADD_SUBDIRECTORY(tpls/rajaperf)
```
Comment (masterleinad) [inline cmake/kokkos_tribits.cmake:86]: Please avoid using tabs.
```suggestion
    ADD_SUBDIRECTORY(tpls/rajaperf)
```

File: core/perf_test/CMakeLists.txt [hunk 5]
```diff
@@ -99,6 +99,17 @@ KOKKOS_ADD_EXECUTABLE_AND_TEST(
   CATEGORIES PERFORMANCE
 )
 ENDIF()
+# AJP, DZP for tracked testing
```
Comment (masterleinad) [inline core/perf_test/CMakeLists.txt:102]:
```suggestion
# Tests using RajaPerfSuite
```
?

File: core/perf_test/ViewAllocationPerfTest.hpp [hunk 6]
```diff
@@ -0,0 +1,34 @@
+#include <Kokkos_Core.hpp>
+#include <common/RAJAPerfSuite.hpp>
+#include <common/Executor.hpp>
+#include <common/QuickKernelBase.hpp>
+
+namespace test {
+
+        void BuildViewAllocationTest(rajaperf::Executor& exec) {
+                exec.registerKernel("VIEW", rajaperf::make_kernel_base("ViewAllocationTest",
+										exec.getRunParams(),
```
Comment (masterleinad) [inline core/perf_test/ViewAllocationPerfTest.hpp:10]: I would be surprised if the tabs here pass `clang-format`.

File: core/perf_test/ViewAllocationPerfTest.hpp [hunk 7]
```diff
@@ -0,0 +1,34 @@
+#include <Kokkos_Core.hpp>
```
Comment (masterleinad) [inline core/perf_test/ViewAllocationPerfTest.hpp:1]: Can you please elaborate on what `core/perf_test/ViewAllocationPerfTest.hpp ` tests and how?

File: core/perf_test/ViewAllocationPerfTest.hpp [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+
+
+
+
+
+        }
+
+
+
```
Comment (masterleinad) [inline core/perf_test/ViewAllocationPerfTest.hpp:33]:
```suggestion
        }
```

**Review by masterleinad — CHANGES_REQUESTED** [review]
> We have created an initial implementation of the perf testing framework for Kokkos. CAVEAT: this does not represent a final, polished state for the implementation, but rather a first draft to communicate and discuss the concept, and elicit feedback.

Did you intend to create this pull request as a draft then?

**Comment by masterleinad** [comment]
>  [@ajpowelsnl](/ajpowelsnl) [ajpowelsnl](/ajpowelsnl) closed this [2 minutes ago](#event-5070538126)

Did you intend to close this?

**Comment by masterleinad** [comment]
You will need to fix the indentation to make `clang-format-8` happy. (One way to do so is running `./scripts/apply-clang-format`).

**Comment by ajpowelsnl** [comment]
@masterleinad -- thanks for the suggestion.  Can I do that in the PR, or should I do it in my project?

**Comment by masterleinad** [comment]
@ajpowelsnl Fix it in this pull request/this branch.

**Comment by masterleinad** [comment]
@ajpowelsnl Any news here?

**Comment by crtrott** [comment]
We are taking a different approach with google bench

**Pull request closed without merging**